client/rpc/debug/src/lib.rs

```diff
@@ -517,13 +542,58 @@ where
 								.ok_or("Trace result is empty.")
 								.map_err(|e| internal_err(format!("{:?}", e)))?
 								.into_iter()
-								.map(|mut trace| {
-									if let Some(transaction_hash) =
-										eth_transactions_by_index.get(&trace.tx_position)
+								.filter_map(|mut trace: BlockTransactionTrace| {
+									if let Some(EthTxPartial {
+										transaction_hash,
+										from,
+										to,
+									}) = eth_transactions_by_index
+										.get(&(trace.tx_position - tx_position_offset))
 									{
-										trace.tx_hash = *transaction_hash;
+										// verify that the trace matches the ethereum transaction
+										let (trace_from, trace_to) = match trace.result {
+											TransactionTrace::Raw { .. } => {
+												(Default::default(), None)
+											}
+											TransactionTrace::CallList(_) => {
+												(Default::default(), None)
+											}
+											TransactionTrace::CallListNested(ref call) => {
+												match call {
+													single::Call::Blockscout(_) => {
+														(Default::default(), None)
+													}
+													single::Call::CallTracer(call) => (
+														call.from,
+														match call.inner {
+															CallTracerInner::Call {
+																to, ..
+															} => Some(to),
+															CallTracerInner::Create { .. } => None,
+															CallTracerInner::SelfDestruct {
+																..
+															} => None,
+														},
+													),
+												}
+											}
+										};
+										if trace_from == *from && trace_to == *to {
```

Comment:
I do not think this will be enough to cover every case, it is possible that a eth tx originated from [polkadotXcm.transfer_assets](https://moonbeam.subscan.io/block/10358165) has the same `from` and `to` as an eth tx from `eth_sendRawTransaction`.

Reply:
I already mentioned this in the PR description:

> Theoretically, it might be possible to have a buggy transaction and a successful eth‑xcm transaction in a row that share the same caller (i.e., the from field), but in practice this never happens and this fix only applies to old blocks.

A proper fix would require updating all past tracing runtimes, but that’s overkill just to correct a few invalid historical traces. We’ve already spent too much time on this fix.

I prefer to merge this as‑is. If we still encounter buggy traces later, we can hardcode the definitive list of problematic transactions in the client after the runtime upgrade.